games/guess-it/src/lib.rs

```diff
@@ -0,0 +1,287 @@
+    pub bet: u128,
+    pub players: HashMap<String, Player>,
+    pub winner: String,
+    pub last_roll: u128,
+}
+impl fmt::Debug for GameSerialized {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        let formatted = serde_json::to_string(&self).unwrap().replace("\\", "");
+        write!(f, "{}", formatted)
+    }
+}
+impl GameSerialized {
+    pub fn empty(players: Option<HashMap<String, Player>>) -> Self {
+        Self {
+            name: "".to_string(),
+            state: State::AcceptingChallenger,
+            bet: 0u128,
+            players: players.unwrap_or(HashMap::new()),
+            winner: NonFungibleKey::from(0).to_string(),
+            last_roll: 0u128,
+        }
+    }
+}
+
+#[derive(Debug, TypeId, Describe, Decode, Encode, Clone, Serialize, Deserialize)]
+pub struct Player {
+    pub guess: u128,
+}
+impl Player {
+    pub fn empty() -> Player {
+        return Self {
+            guess: 0u128,
+        }
+    }
+
+    pub fn get_guess(&self) -> u128 { self.guess }
+}
+
+#[derive(TypeId, Describe, Decode, Encode)]
+#[derive(NonFungibleData)]
+struct PlayerNft {
+    name: String,
+}
+
+blueprint! {
+    struct GuessIt {
+        bank: Vault,
+        badges: Vault,
+        badge_ref: ResourceDef,
+        game: Game,
+    }
+
+    impl GuessIt {
+        pub fn create(name: String, bet: u128) -> Component {
+            let player_badges: Bucket = ResourceBuilder::new_non_fungible()
+                .metadata("name", "Guess-It Game")
+                .metadata("symbol", "GIG")
+                .initial_supply_non_fungible([
+                    (
+                        NonFungibleKey::from(Uuid::generate()),
+                        PlayerNft {name: "Player 1".to_string()},
+                    ),
+                    (
+                        NonFungibleKey::from(Uuid::generate()),
+                        PlayerNft {name: "Player 2".to_string()},
+                    ),
+                ]);
+            let player_badge_def = player_badges.resource_def();
+
+            Self {
+                bank: Vault::new(RADIX_TOKEN),
+                badges: Vault::with_bucket(player_badges),
+                badge_ref: player_badge_def,
+                game: Game::new(None, name, bet),
+            }.instantiate()
+        }
+
+        pub fn join(&mut self, mut payment: Bucket) -> (Bucket, Bucket, String) {
+            assert_eq!(self.game.state, State::AcceptingChallenger, "Not accepting challengers!");
+            assert!(self.game.players.len() <= 2, "Too many players!");
+
+            let player: Player = Player::empty();
+            let badge: Bucket = self.badges.take(Decimal::one());
+            let nft_id: NonFungibleKey = badge.get_non_fungible_key();
+            // Secure the funds
+            self.bank.put(payment.take(Decimal::from(self.game.bet as i128)));
+            // Add the player's details
+            self.game.players.insert(nft_id.clone(), player.clone());
```

Comment:
Is `clone()` needed here?

Reply:
On: 224 - 228: a move occurs and I get the error ` move occurs because `nft_id` has type `scrypto::prelude::NonFungibleKey`, which does not implement the `Copy` trait`

It seems I do not need the clone for the player tho.

Will update this in upcoming commit